src/types/osstr.rs

```diff
@@ -0,0 +1,191 @@
+            let len = wstr.len() as ffi::Py_ssize_t;
+            unsafe {
+                // HELP: is there some better thing to do here than panic? Some last resort thing to
+                // try?
+                py.from_owned_ptr_or_err::<PyString>(ffi::PyUnicode_FromWideChar(ptr, len))
+                    .expect("OsStr conversion to Python string via FromWideChar failed")
+                    .into()
+            }
+        }
+    }
+}
+
+// It's not possible to implement FromPyObject for OsStr on Windows, because Python's wchar
+// representation can't be directly mapped to OsStr's one
+// HELP: should this implementation be here at all? It may be confusing for PyO3 users when a trait
+// is implemented only on some platforms
+#[cfg(not(windows))]
+impl<'a> FromPyObject<'a> for &'a OsStr {
+    fn extract(ob: &'a PyAny) -> PyResult<Self> {
+        let pystring = <PyString as PyTryFrom>::try_from(ob)?; // Cast PyAny to PyString
+
+        let fs_encoded_bytes: &crate::types::PyBytes = unsafe {
+            ob.py()
+                .from_borrowed_ptr(ffi::PyUnicode_EncodeFSDefault(pystring.as_ptr()))
+        };
+
+        Ok(std::os::unix::ffi::OsStrExt::from_bytes(
+            fs_encoded_bytes.as_bytes(),
+        ))
+    }
+}
+
+impl FromPyObject<'_> for OsString {
+    fn extract(ob: &PyAny) -> PyResult<Self> {
+        #[cfg(not(windows))]
+        {
+            let os_str: &OsStr = ob.extract()?;
+            Ok(os_str.to_os_string())
+        }
+
+        #[cfg(windows)]
+        {
+            let pystring = <PyString as PyTryFrom>::try_from(ob)?; // Cast PyAny to PyString
+
+            // Take the quick and easy shortcut if UTF-8
+            if let Ok(utf8_string) = pystring.to_str() {
+                return Ok(utf8_string.to_owned().into());
+            }
+
+            // The Windows way involves two allocations+copies, unfortunately.
+            // First, we have to convert the opaque PyString into a wide char buffer. Python exposes
+            // no methods to get a view into its internal buffers, so this step can only be done via
+            // copying.
+            // Second, we have to convert the wide char buffer into a OsString. Rust provides no
+            // way to cheaply convert an owned buffer into an OsString; it only exposes a method
+            // involving a copy and an allocation.
+
+            // Get an owned allocated wide char buffer from PyString, which we have to deallocate
+            // ourselves
+            let mut size: ffi::Py_ssize_t = 0;
+            let buffer_ptr =
+                unsafe { ffi::PyUnicode_AsWideCharString(pystring.as_ptr(), &mut size) };
+
+            // Copy wide char buffer into OsString
+            let os_string = std::os::windows::ffi::OsStringExt::from_wide(unsafe {
+                std::slice::from_raw_parts(buffer_ptr, size as usize)
+            });
+
+            unsafe {
+                ffi::PyMem_Free(buffer_ptr as *mut libc::c_void);
```

Comment:
NB we will leak memory if `OsStringExt::from_wide` panics - easiest fix is probably to wrap `buffer_ptr` in a struct which calls `PyMem_Free` in a `Drop` impl. NB maybe better to use `Vec` - see comment on line 105.